### DISCUSSION TITLE
New graph data structure

### DISCUSSION THREAD

**Discussion by angriman** [networkit/networkit#691] [Ideas]
From our internal meeting we discussed to convert the graph data structure from an adjacency list to a CSR matrix. It is my understanding that this implies two fundamental changes to all the classes using in-edges iterators (e.g., `forInNeighborsOf`):
- they will take the transposed graph as an additional constructor parameter
- in-edges iterators will be converted to out-edges iterators over the transposed graph

IMHO this would change too much NetworKit APIs. My suggestion is to:
- store the transposed CSR matrix within the `Graph` class
- rewrite in-edge iterators to use the transposed matrix
- add an helper method to `Graph` to allocate the transposed matrix when necessary

**Comment by manpen** [comment]
As far as I'm concerned the main reason for not storing the transposed representation (independently of whether we're doing it as CSR, Adj matrix,...) is that it almost always eliminates any locality an algorithm might have. This is especially an issue when working in the parallel context, where we might be able to restrict threads to certain disjoint nodes, allowing for parallel changes of the Graph. This is impossible if we always have to fix the arbitrary back-edges.

So another approach would be to store the forward edges in an instance, and if a transposed variant is required, we can compute it. Then have a thin wrapper that take the reference to both objects and simulates the old API.

**Comment by avdgrinten** [comment]
For reference, this is the set of affected files:
```
networkit/cpp/io/NetworkitBinaryWriter.cpp
networkit/cpp/distance/BidirectionalDijkstra.cpp
networkit/cpp/distance/ReverseBFS.cpp
networkit/cpp/distance/AffectedNodes.cpp
networkit/cpp/distance/DynBFS.cpp
networkit/cpp/distance/DynDijkstra.cpp
networkit/cpp/distance/DynAPSP.cpp
networkit/cpp/distance/AllSimplePaths.cpp
networkit/cpp/components/DynWeaklyConnectedComponents.cpp
networkit/cpp/components/WeaklyConnectedComponents.cpp
networkit/cpp/generators/DynamicForestFireGenerator.cpp
networkit/cpp/centrality/KatzCentrality.cpp
networkit/cpp/centrality/GroupDegree.cpp
networkit/cpp/centrality/DynBetweenness.cpp
networkit/cpp/centrality/ApproxCloseness.cpp
networkit/cpp/centrality/EigenvectorCentrality.cpp
networkit/cpp/centrality/CoreDecomposition.cpp
networkit/cpp/centrality/PageRank.cpp
networkit/cpp/centrality/DynApproxBetweenness.cpp
networkit/cpp/centrality/KadabraBetweenness.cpp
networkit/cpp/centrality/DynBetweennessOneNode.cpp
networkit/cpp/centrality/DynKatzCentrality.cpp
```

**Comment by avdgrinten** [comment]
If we keep the in-edges and add a flag whether they should be stored or not, I see two options for algorithms that mutate graphs: (1) add a method `rebuildTranspose(Unsafe)` that rebuilds the transposed graph from scratch, or (2) use a class `GraphMutator` similarly to `GraphBuilder` that records the desired updates in a queue and uses an efficient parallel algorithm to update both the graph and the transpose at the same time.

**Comment by hmeyerhenke** [comment]
When making such a fundamental change, please ensure that we don't make changes to the graph (vertex/edge insertions/deletions) prohibitively expensive. I am not sure if I mentioned this concern already in our meeting.

**Comment by clstaudt** [comment]
If such a fundamental change is really on the table, here's something I have been thinking about: What about building on existing, well-supported data structures? 

https://docs.scipy.org/doc/scipy/reference/sparse.html
https://www.tensorflow.org/api_docs/python/tf/sparse/SparseTensor

With respect to making those extensible, perhaps get inspired by how `pandas` solves this:
https://www.youtube.com/watch?v=F37fV0uFf60

Why did we go for our own data structure in the first place? I think one reason is that we are being taught how to do that at university, but that may not be the economical way to do it.

**Comment by avdgrinten** [comment]
The SciPy and TensorFlow data structures are also nothing fundamentally different from CSR. The challenge / research objective is making it dynamic, not implementing the CSR.

**Comment by clstaudt** [comment]
Exactly my point. And perhaps something like pandas has basically already solved the challenge - not sure without looking into the details. Pandas is semi-dynamic, i.e. dynamic with performance penalties.

**Comment by hmeyerhenke** [comment]
Responding to clstaudt:
> Why did we go for our own data structure in the first place? I think one reason is that we are being taught how to do that at university, but that may not be the economical way to do it.

No, because existing data structures did/do not offer efficient support for dynamic updates.

BTW: Some people from the linear algebra community are now actively approaching the graph analytics community to see how to implement dynamic graph/matrix data structures...

**Comment by clstaudt** [comment]
Let me rephrase: At that time we did not research whether existing implementations, e.g. from the NumPy/SciPy code base, could be adapted for the purpose.  If I remember correctly it was not considered at all, since implementing new data structures from scratch seemed like the obvious choice.  A custom implementation may still be the best choice, but it should come after evaluation of the standard tools. Another interesting benefit to be researched would be that reuse of data structures could enable better integration with the SciPy (or maybe even TensorFlow) ecosystem, therefore adoption and reuse of other scientific computing code. A custom data structure may have the best performance for dynamic updates, but that's not the only relevant dimension. If nothing else, an evaluation of the tradeoffs would be a strong point of a paper.

**Comment by larsgottesbueren** [comment]
I found this paper http://supertech.csail.mit.edu/papers/WheatmanXu18.pdf for a "dynamic" CSR. What I've understood from a quick glance is 1) they leave empty slots in the arrays for fast insertions, which sounds like the most reasonable approach 2) slowdown compared to CSR on a number of benchmarks is roughly a factor of 2.

**Comment by hmeyerhenke** [comment]
Such a "dynamic CSR" is not really new, but thanks for the pointer. How to respond to very irregular inputs is maybe not so obvious and established. Integration into/with other relevant ecosystems should be a consideration, yes.
